Enable CrossPlatform Builds and CI

MathewBensonCode opened the pull request:
# Changes made to enable cross platform builds

- Upgrade submodule dependencies to newer versions as upstream generally have fixed issues with them. For example assimp -Werror issues and GLM requiring old CMake Flags
- Rename variables that have the same name as the naming type as this wasn't working with GCC.
- Replicate Linux CI workflow using Ninja Generator for Linux
- Remove hard requirement for gcc11 and let the default compiler on the system to be used

JeanPhilippeKernel commented on Scripts/BuildEngine.ps1:
```diff
@@ -131,28 +116,26 @@ function Build([string]$configuration, [int]$VsVersion , [bool]$runBuild) {
     $submoduleCMakeOptions = @{
         'ENTT'      = @("-DENTT_INCLUDE_HEADERS=ON")
-        'SPDLOG'    = @("-DSPDLOG_BUILD_SHARED=OFF", "-DBUILD_STATIC_LIBS=ON", "-DSPDLOG_FMT_EXTERNAL=ON", "-DSPDLOG_FMT_EXTERNAL_HO=OFF");
-        'GLFW '     = @("-DGLFW_BUILD_DOCS=OFF", "-DGLFW_BUILD_EXAMPLES=OFF", "-DGLFW_INSTALL=OFF");
```
any reason why removing "-DGLFW_BUILD_EXAMPLES=OFF"  ?

MathewBensonCode replied:
I read the upstream cmake file and found it was being set by default so I was removing the redundant option.

https://github.com/glfw/glfw/blob/master/CMakeLists.txt

JeanPhilippeKernel commented on Scripts/PostBuild.ps1:
```diff
@@ -43,11 +43,7 @@ $ErrorActionPreference = "Stop"
-}  Else {
-    [IO.Path]::Combine($RepoRoot, "Result.$SystemName.x64.$Configurations")
-}
+[string]$OuputBuildDirectory =  [IO.Path]::Combine($RepoRoot, "../builds/RendererEngine")
```
Not sure this change 
the project has been configured to redirect the build output into a particular directory based on the platform - arch - and config

MathewBensonCode replied:
will revert this, was left there accidentally.

JeanPhilippeKernel commented on Scripts/Shared.ps1:
```diff
@@ -185,8 +185,6 @@ function Find-GlslangValidator () {
 
 function Find-ClangFormat () {
     $repoConfiguration = Get-RepositoryConfiguration
-    $LLVMMinimumVersion = $repoConfiguration.Requirements.LLVM.Version
-    $LLVMMaximumVersion = $repoConfiguration.Requirements.LLVM.MaximumVersion
```
please revert -  LLVM has the tendency to introduce breaking change between each version -  and to avoid that we enforce sticking to specific version - which is configure in `repoConfiguration.json file`

MathewBensonCode replied:
I think we can give this a bit more thought as limiting the version will limit portability and flexibility. 

I was envisioning the ability to have the system build on a wide variety of configurations.

Perhaps testing can be done on the current stable version and have that set as the maximum.

JeanPhilippeKernel replied:
I believe we are already on the latest stable version: 
the config for external tool like llvm here is: 
https://github.com/JeanPhilippeKernel/RendererEngine/blob/develop/repoConfiguration.json

MathewBensonCode replied:
I see. I made some changes to the CI workflow to ensure the linux build can find and use clang-format 20 on ubuntu and some other distributions as well by adding options tp the searched paths.

JeanPhilippeKernel commented on Scripts/Shared.ps1:
```diff
@@ -202,19 +200,16 @@ function Find-ClangFormat () {
         }
     }
 
-    throw "Failed to find clang-format min $LLVMMinimumVersion max $LLVMMaximumVersion. Tried: " + ($candidates -join ', ')
```
Same here - see my comment about the LLVM breaking change

JeanPhilippeKernel commented on Scripts/Shared.ps1:
```diff
@@ -202,19 +200,16 @@ function Find-ClangFormat () {
 }
 
 function Setup-ShaderCCompilerTool () {
+    if($IsLinux) {
+        return
+    }
```
I believe ShaderC compiler has a version for linux -- @jnyfah can you check that ?

jnyfah replied:
yes it does

MathewBensonCode replied:
I'll look into this. At a glance I saw that the dependency was being pulled in from an archive which I assumed to be yours so I was testing with the system installed version.

JeanPhilippeKernel commented on ZEngine/CMakeLists.txt:
```diff
@@ -6,18 +6,5 @@ project (ZEngine
 	LANGUAGES CXX C
 )
 
-# Compiler verification in Linux
-#
-if (${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
-	if ((NOT ${CMAKE_CXX_COMPILER_VERSION} MATCHES "11.*.*") AND (${CMAKE_CXX_COMPILER_ID} STREQUAL "Clang"))
-		message(FATAL_ERROR "CLANG compiler version should be 11.*.* !")
-	elseif ((NOT ${CMAKE_CXX_COMPILER_VERSION} MATCHES "11.*.*") AND (${CMAKE_CXX_COMPILER_ID} STREQUAL "GNU"))
-		message(FATAL_ERROR "GNU compiler version should be 11.*.* !")
-	endif ()
-endif ()
```
at the time I added this compiler verification - GCC 11 was the version that has a non-buggy support for coroutine?
do you ensure that all newer versions are good now? same for Clang

MathewBensonCode replied:
I think its safe to say that this is ok. The CI version of ubuntu uses version 13, and I have tested 15 and 16(trunk).

I think it would only break with older compilers which should be quite rare.

JeanPhilippeKernel replied:
as you said, it will break with older versions -  let's enforce that by telling the developer to upgrade its compiler version.

The project needs C++20 and coroutine support
for Clang  it should be minimum v11 and higher
for GCC it should be minimum v10 or 11  and higher

MathewBensonCode replied:
This is actually being enforced by the CMAKE_CXX_STANDARD variable. If the compiler does not support c++20,  then the process ends here. CMake is aware of the versions and support so we do not have to do it manually.

https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html

JeanPhilippeKernel commented on ZEngine/ZEngine/Rendering/Renderers/GraphicRenderer.cpp:
```diff
@@ -15,7 +15,7 @@
 #define STB_IMAGE_WRITE_IMPLEMENTATION
 #define STB_IMAGE_RESIZE_IMPLEMENTATION
-#include <stb/stb_image_resize.h>
+#include <stb/deprecated/stb_image_resize.h>
```
any reason ?

MathewBensonCode replied:
the upstream project moved this header to the deprecated folder as there is a newer version that would need a code rewrite to match.

JeanPhilippeKernel commented on ZEngine/ZEngine/Windows/Inputs/KeyCode.h:
```diff
@@ -136,4 +137,9 @@ namespace ZEngine::Windows::Inputs
 
     } GlfwKey;
+
+    inline auto format_as(GlfwKey f)
+    {
+        return fmt::underlying(f);
+    }
```
??

MathewBensonCode replied:
There was a compilation error when trying to pass the enum glfwkey to the fmt lib. This is to tell fmt to use the underlying type of the enum as the formating type.

JeanPhilippeKernel commented on ZEngine/ZEngine/CMakeLists.txt:
```diff
@@ -67,16 +58,9 @@ target_compile_definitions (zEngineLib
 
-if(${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
-	target_compile_definitions (zEngineLib PUBLIC VK_USE_PLATFORM_WIN32_KHR)
-endif()
```
any reason ??

JeanPhilippeKernel commented on ZEngine/ZEngine/CMakeLists.txt:
```diff
@@ -67,16 +58,9 @@ target_compile_definitions (zEngineLib
-
-if (${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
-	target_link_libraries(zEngineLib PRIVATE stdc++fs)
-endif ()
```
same any reason ?

JeanPhilippeKernel commented on __externals/externals.cmake:
```diff
@@ -1,60 +1,80 @@
-add_library(imported::External_editorLibs INTERFACE IMPORTED)
+if (${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
+    target_link_libraries(ImGuizmo PUBLIC stdc++fs)
+endif ()
```
No, ENABLE_VULKAN_VALIDATION_LAYER, VK_USE_PLATFORM_WIN32_KHR and stdc++fs are needed by zEngine lib , not ImGuizmo.  please revert

MathewBensonCode replied:
I had meant to put this in the imgui target as it is the one that pulls in vulkan.

The idea is that imgui pulls in glfw and vulkan, with their definitions. Then Imguizmo pulls in imgui and finally ZEngine pulls in ImGuizmo.

JeanPhilippeKernel commented on __externals/externals.cmake:
```diff
@@ -1,60 +1,80 @@
+add_subdirectory (${EXTERNAL_DIR}/CLI11)
+add_subdirectory (${EXTERNAL_DIR}/nlohmann_json)
+
+add_library (ZEngineLibs INTERFACE)
+add_library(TetragramaLibs INTERFACE)
```
this naming is confusing - the engine lib name itself is zEnginelib

MathewBensonCode replied:
I agree. Perhaps I could rename them as ZEngineLibDependencies or something similar. 

In the same vein we should consider using a different name for the imported::*** versions.

JeanPhilippeKernel commented on .clang-format:
```diff
@@ -120,7 +120,7 @@ PenaltyBreakTemplateDeclaration: 100
 PenaltyReturnTypeOnItsOwnLine: 300
 PointerAlignment: Left
 ReflowComments: true
-SortIncludes: true
```
any reason ?

MathewBensonCode replied:
There was a compilation issue that would come up after running clang-format in ImGUIRenderer.cpp where the ImGuizmo header when included before the imgui header would break as the imgui headers need to come in first.

JeanPhilippeKernel replied:
we can simply put that header in a formatting off section `// clang-format off ..... // clang-format on`
see example of usage here : https://github.com/JeanPhilippeKernel/RendererEngine/blob/develop/ZEngine/ZEngine/Hardwares/VulkanDevice.h

JeanPhilippeKernel commented on __externals/imgui/CMakeLists.txt:
```diff
@@ -14,16 +14,20 @@ list (APPEND IMGUI_SOURCE
 target_include_directories (imgui
     PUBLIC
         ./src
-        ./src/backends
```
we don't need src/backend anymore ?

MathewBensonCode replied:
The backends directory will be specified in the #include directive in the source files.

jnyfah commented:
@MathewBensonCode I think to make it easier you can break this PR into smaller PR's that focus on one thing at a time 
maybe first work on 
1. adding linux support and ensuring that the panzerfaust/launcher and the editor works correctly 
2. Once the Linux support PR is merged you can work on adding linux workflows
3. finally fix the submodule 

that way its not overwhelming and things are fixed one at a time 
let me know if you have any questions or issues

JeanPhilippeKernel commented on .github/workflows/job-deploy-linux.yml:
```diff
@@ -0,0 +1,29 @@
+    steps:
+      - uses: actions/download-artifact@v4
+        with:
+          path: Result.Darwin.x64.${{ inputs.configuration }}
```
```suggestion
          path: Result.Linux.x64.${{ inputs.configuration }}
```

JeanPhilippeKernel commented on .github/workflows/linux-build.yml:
```diff
@@ -1,59 +1,37 @@
-# @JeanPhilippeKernel : Disabled because we only support Windows as platform for now
+# @JeanPhilippeKernel 
 #
```
we can drop this comment

MathewBensonCode commented:
I have tried to capture the requested changes in a single commit which we can now look into splitting up into smaller PRs as necessary.

I found it a little tricky to split up because the changes were related but we can now discuss this further and you can let me know what you have in mind.

JeanPhilippeKernel commented on ZEngine/ZEngine/Windows/Inputs/KeyCode.h:
```diff
@@ -136,4 +137,10 @@ namespace ZEngine::Windows::Inputs
 
     } GlfwKey;
+
+    inline std::ostream& operator<<(std::ostream& stream, GlfwKeyCode f)
+    {
+        stream << static_cast<int32_t>(f);
+        return stream;
+    }
```
any reason for this overload ?

MathewBensonCode replied:
some compilers or versions of fmt have trouble formating enums. On newer versions there is a `format_as` function, but as We are not upgrading the submodules the `ostream <<` is the alternative.

https://github.com/fmtlib/fmt/issues/391

JeanPhilippeKernel replied:
interesting - can you help creating an issue to track the upgrade of fmtlib

JeanPhilippeKernel commented on Scripts/RunTests.ps1:
```diff
@@ -79,11 +81,11 @@ function RunTests {
 # Run tests for each configuration
 foreach ($config in $Configurations) {
-    RunTests -Configuration $config
+RunTests -Configuration $config
```
revert the space

JeanPhilippeKernel commented on Scripts/Shared.ps1:
```diff
@@ -189,14 +189,18 @@ function Find-ClangFormat () {
     $LLVMMaximumVersion = $repoConfiguration.Requirements.LLVM.MaximumVersion
 
     $candidates = @(
-        'clang-format'
```
revert - it should be the first candidate before looking the specific path

MathewBensonCode replied:
Doing it this way is important where clang-format version 20 is not the default, for example in the CI build where we are specifically installing it as well as on a system like mine where the path always selects the newer version first. If some other system specific version or path comes up in testing it would be easy to just add that to the search options

JeanPhilippeKernel replied:
I see - alright then

JeanPhilippeKernel commented on Scripts/BuildEngine.ps1:
```diff
@@ -171,11 +169,7 @@ function Build([string]$configuration, [int]$VsVersion , [bool]$runBuild) {
-            # Set Linux build compiler
-            $env:CC = '/usr/bin/gcc-11'
-            $env:CXX = '/usr/bin/g++-11'
+            $cMakeGenerator = "-G `"Ninja`""
```
we want to update the README to provide installation steps for Ninja  on linux

Ideally we might want to automate that installation so devs wont need to worry about it.  we can do that in follow-up PR - please create an issue to track it

MathewBensonCode replied:
done

MathewBensonCode committed:
Enable Linux Build and Continuous Integration

- Renamed variables that have the same name as the naming type as this wasn't compiling on GCC.
- Replicated Linux CI workflow using Ninja Generator for Linux
- Removed hard requirement for gcc and clang version 11 and use CMAKE_CXX_STANDARD = 20
  enforce this.
- Enabled Linux Output Paths in Panzerfaust.csproj
- Add Search Paths for clang-format specific to possible location in
  different linux versions.(list may be incomplete for other
  distributions)

jnyfah reviewed (changes requested):
i dont know why i cant comment on the file, but `shared.ps1` add linux to the `GlslangValidatorCandidates` line 161

`
    $GlslangValidatorCandidates = @(
        'glslangValidator'

        if ($IsWindows) {
            Join-Path -Path $shaderCCompilerPath -ChildPath "\bin\glslangValidator.exe" # On Windows, the pipeline build might pick up this option...
        }
        if ($IsMacOS) {
            Join-Path -Path $shaderCCompilerPath -ChildPath "\bin\glslangValidator" # On macOS, the pipeline build might pick up this option...
        }
        if ($IsLinux) {
           Join-Path -Path $shaderCCompilerPath -ChildPath "\bin\glslangValidator"
        }
    )
`

MathewBensonCode committed:
Add glslangValidator path for linux

jnyfah reviewed (changes requested):
everything looks good to me, just make the changes on the glslangcandidate
